# 4.8 Grouping Data & Aggregating Variables Task

In [1]:
#Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
#Importing dataframe

df_ords_prods_merge = pd.read_pickle(r'/Users/dorinasalam/Desktop/08-2023 Instacart Basket Analysis/02 Data/Prepared Data/orders_products_flagged.pkl')

In [3]:
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### The results for the entire dataframe do not differ much from the subset results. They have minor differences, but nothing that looks visually significant.  

In [4]:
df_ords_prods_merge.groupby('department_id').agg({'orders_day_of_week': ['mean']})

orders_day_of_week
                            mean
department_id                   
1                       2.790508
2                       2.851368
3                       2.743055
4                       2.683672
5                       3.163153
6                       2.691937
7                       2.807199
8                       2.934893
9                       2.690505
10                      2.759061
11                      2.816660
12                      2.718082
13                      2.810497
14                      2.731141
15                      2.707202
16                      2.739654
17                      2.835449
18                      2.711004
19                      2.751490
20                      2.705505
21                      2.800853

### All departments show the same mean day of the week = 2 (Monday)

# Identifying customer spending habits:

In [5]:
#Determining spending habits amongst loyaltly users.

df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## The mean amount spent by a loyal customer is slightly less than the regular or new customer. 
### Unknown whether it is significantly different yet.

In [6]:
df_ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   
5   1701441      777            16                   1                  7   
6   1871483      825             3                   2                 14   
7   1290456      910            12                   3                 10   
8    369558     1052            10                   1                 20   
9    589712     1052            15                   1                 12   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   
5                    26.0                  7          0   both   
6                    30.0                  2          0   both   
7                    30.0                  1          0   both   
8                    19.0                  1          0   both   
9                    15.0                  2          1   both   

     price_range_loc     busiest_day busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy           Most orders         32   
1  Mid-range product  Regularly busy        Average orders         32   
2  Mid-range product     Busiest day        Average orders          5   
3  Mid-range product   Slightly busy           Most orders          3   
4  Mid-range product      Least busy        Average orders          3   
5  Mid-range product       Very busy        Average orders         26   
6  Mid-range product  Regularly busy           Most orders          9   
7  Mid-range product   Slightly busy           Most orders         12   
8  Mid-range product       Very busy        Average orders         20   
9  Mid-range product       Very busy           Most orders         20   

       loyalty_flag  
0  Regular customer  
1  Regular customer  
2      New customer  
3      New customer  
4      New customer  
5  Regular customer  
6      New customer  
7  Regular customer  
8  Regular customer  
9  Regular customer

# Identifying different types of spenders:

In [7]:
#Creating new column for mean_price per customer.

df_ords_prods_merge['mean_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [8]:
df_ords_prods_merge[['mean_price','user_id']].head(40)

mean_price  user_id
0     6.935811      138
1     6.935811      138
2     7.930208      709
3     4.972414      764
4     4.972414      764
5     6.935398      777
6     5.957576      825
7     6.680000      910
8     7.162500     1052
9     7.162500     1052
10    5.813636     1379
11    5.081633     1480
12    5.081633     1480
13    5.081633     1480
14    6.803797     1494
15    6.803797     1494
16    6.803797     1494
17    6.607122     1540
18    6.607122     1540
19    6.607122     1540
20    6.607122     1540
21    6.607122     1540
22    6.607122     1540
23    6.607122     1540
24    6.607122     1540
25    6.607122     1540
26    6.607122     1540
27    6.607122     1540
28    6.607122     1540
29    6.607122     1540
30    6.607122     1540
31    6.607122     1540
32    6.607122     1540
33    6.607122     1540
34    7.123529     1598
35    7.123529     1598
36    7.123529     1598
37    4.853521     1647
38    4.853521     1647
39    4.853521     1647

In [9]:
#Creating flag for high and low spenders

df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] < 10, 'spender_flag'] = 'Low spender'
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] >= 10, 'spender_flag'] = 'High spender'

In [10]:
df_ords_prods_merge['spender_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spender_flag, dtype: int64

### There are 31,770,614 low spenders and 634,245 high spenders.

# Identifying frequency of customers:

In [11]:
#Creating new column for median frequency per customer.

df_ords_prods_merge['median_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [12]:
#Creating flag for frequency of customers

df_ords_prods_merge.loc[df_ords_prods_merge['median_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['median_frequency'] > 10) & (df_ords_prods_merge['median_frequency'] <= 20), 'frequency_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['median_frequency'] < 10, 'frequency_flag'] = 'Frequent customer'

In [13]:
df_ords_prods_merge['frequency_flag'].value_counts()

Frequent customer        20287988
Regular customer          7208564
Non-frequent customer     3636437
Name: frequency_flag, dtype: int64

In [14]:
df_ords_prods_merge.to_pickle(r'/Users/dorinasalam/Desktop/08-2023 Instacart Basket Analysis/02 Data/Prepared Data/orders_products_customer.pkl')